File: 04- April/15- Maximum Value of K Coins From Piles/15- Maximum Value of K Coins From Piles (Ahmed Gamal).js
// Author: Ahmed Gamal

// for this problem we will use dp to find the maximum value of k coins from piles
// dp[i][j]: the maximum value of j coins from piles starting from pile i (i: pile index, j: number of coins)
// and since we can only take coins from the end of the pile, we will use the prefix sum array to find the maximum value of j coins from piles starting from pile i
// because the maximum value of choosing j coins from the end of the pile is the sum of the last j coins in the pile

// we can do this prefix sum inplace by iterating over the piles array and updating the current element with the sum of the current element and the previous element
// then we can iterate over the possible number of coins we can take from the end of the pile and find the maximum value of choosing j coins from the end of the pile

/**
 * @param {number[][]} piles
 * @param {number} k
 * @return {number}
 */
var maxValueOfCoins = function(piles, k) {
    // prefix sum
    piles.forEach(pile => {
        for(let i = 1; i < pile.length; i++) {
            pile[i] += pile[i - 1];
        }
    });

    // memo: memoization array
    // n: number of piles
    let memo = [];
    const n = piles.length;

    // dp: dp function
    // idx: pile index
    // rem: number of coins remaining
    let dp = (idx, rem) => {
        // if we reached the end of the piles array
        if(idx >= n) {
            // if we took all the coins return 0 else return -1e10 (negative infinity)
            return rem ? -1e10 : 0;
        }

        // if this state is undefined, initialize it
        if(!memo[idx]) {
            memo[idx] = [];
        }
        // if we already calculated this state return the result
        if(memo[idx][rem]) {
            return memo[idx][rem];
        }

        // ret: the maximum value of choosing rem coins from piles starting from pile idx

        // try to take 0 coins from the end of the pile
        let ret = dp(idx + 1, rem);

        // try to take 1..min(rem, piles[idx].length) coins from the end of the pile
        for(let i = 1; i <= rem; i++) {
            // if we reached the end of the pile break
            if(i - 1 >= piles[idx].length) {
                break;
            }
            // update ret with the maximum between ret and the maximum value of choosing rem - i coins from piles starting from pile idx + 1 plus the sum of the last i coins in the pile
            ret = Math.max(ret, piles[idx][i - 1] + dp(idx + 1, rem - i));
        }

        // memoize the result
        memo[idx][rem] = ret;

        // return the result
        return memo[idx][rem];
    };

    // return the maximum value of choosing k coins from piles starting from pile 0
    return dp(0, k);
};

File: 04- April/15- Maximum Value of K Coins From Piles/15- Maximum Value of K Coins From Piles (Ahmed Gamal).cpp
// Author: Ahmed Gamal

// this is the iterative solution for the problem with the same idea of the recursive solution
// for more details about the problem, and solution idea, please check the recursive solution implemented in javascript

class Solution {
public:
    int maxValueOfCoins(vector<vector<int>>& piles, int k) {
        // n piles, k coins
        const int n = int(piles.size());

        // dp[i][j] = maximum value of j coins from the first i piles
        // dp: the previous row
        vector<int> dp(k + 1);


        // for each pile
        for(int i = 0; i < n; i++) {
            // dp_: the current row
            vector<int> dp_(k + 1);

            // for each number of coins remaining
            for(int j = 0; j <= k; j++) {
                // the answer for the current row is the same as the previous row (initially)
                dp_[j] = dp[j];

                // for each number of coins we can take from the current pile
                int sum = 0;
                for(int p = 0; p <= j; p++) {
                    // if we reached the end of the current pile, break
                    if(p > piles[i].size()) {
                        break;
                    }
                    // if we can take p coins from the current pile, and we can take j - p coins from the previous piles
                    dp_[j] = max(dp_[j], dp[j - p] + sum);

                    // add the next coin from the current pile
                    if(p < piles[i].size())
                        sum += piles[i][p];
                }
            }

            // swap the current row with the previous row
            swap(dp, dp_);
        }

        // the answer is the maximum value of k coins from the first n piles
        return dp[k];
    }
};